src/softmax.jl

```diff
@@ -26,49 +32,20 @@ julia> softmax([1, 2, 3])
 
 See also [`logsoftmax`](@ref).
 """
-function softmax(xs::AbstractArray; dims=1)
-    max_ = maximum(xs, dims=dims)
-    exp_ = exp.(xs .- max_)
-    exp_ ./ sum(exp_, dims=dims)
+softmax(x; dims = 1) = softmax!(similar(x), x; dims)
+softmax!(x; dims = 1) = softmax!(x, x; dims)
+function softmax!(out::T, x::T; dims = 1) where {T<:AbstractArray}
```

Comment:
we can relax the constraint of `out` and `x` being of the same type

Reply:
I have updated the constraint for `softmax!` `logsoftmax!`.
For `∇softmax` , I guess `Δ` and `out` should be the same type. Right?

I added only one test case for integer input. is this OK?
```julia
@test softmax(Int[0, 0]) == [0.5, 0.5]
```